File: Daily/645. Set Mismatch/Solution2.py
from typing import List
class Solution:
    def findErrorNums(self, nums: List[int]) -> List[int]:
        '''
            Optimized Solution Using Array or List
            Time Complexity: O(n)
            Space Complexity: O(n)
        '''
        total_Elements, duplicate, missing = len(nums), -1, -1
        frequency = [0] * (total_Elements + 1)
        for element in nums: frequency[element] += 1
        for element in range(1, total_Elements + 1):
            if frequency[element] > 1: duplicate = element
            elif frequency[element] == 0: missing = element
        return [duplicate, missing]

if __name__ == '__main__':
    nums = [int(x) for x in input().split()]; sol = Solution()
    ans = sol.findErrorNums(nums)
    print('Duplicate Number =', ans[0], 'and Missing Number =', ans[1])

File: Daily/645. Set Mismatch/Solution2.cpp
#include <iostream>
#include <vector>
using namespace std;
class Solution {
    public:
        vector<int> findErrorNums(vector<int>& nums) {
            /*
                Optimized Solution using Array or Vector
                Time Complexity = O(n)
                Space Complexity = O(n)
            */
            const int total_Elements = nums.size();
            int duplicate = -1, missing = -1;
            vector<int> frequency(total_Elements+1);
            for(const int& element: nums) frequency[element]++;
            for(int element = 1; element <= total_Elements; ++element) {
                if(frequency[element] > 1) duplicate = element;
                else if(frequency[element] == 0) missing = element;
            }
            return {duplicate, missing};
        }
};
int main() {
    int n; cin >> n; vector<int> nums(n); Solution sol;
    for(int index = 0; index < n; ++index) cin >> nums[index];
    vector<int> ans = sol.findErrorNums(nums);
    cout << "Duplicate Number = " << ans[0] 
        << " and Missing Number = " << ans[1] << endl;
}